File: leetcode/python/0169. Majority Element.py
# Time complexity: O(N)
# Space complexity: O(N)

class Solution(object):
    def majorityElement(self, nums):
        freq_table = {}

        for num in nums:
            if num in freq_table:
                freq_table[num] += 1
            else:
                freq_table[num] = 1
        
        req = len(nums) / 2

        for num, count in freq_table.items():
            if count > req:
                return num
        return -1
        

File: leetcode/java/0169. Majority Element.java
package leetcode.java;

import java.util.HashMap;

// Time complexity: O(N)
// Space complexity: O(N)

class Solution {
    public int majorityElement(int[] nums) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int req = nums.length / 2;

        for (int num : nums) {
            if (map.containsKey(num)) {
                map.put(num, map.get(num)+1);
            } else {
                map.put(num, 1);
            }
        }

        for (int num : map.keySet()) {
            if (map.get(num) > req) {
                return num;
            }
        }
        return -1;
    }
}